### PR TITLE
Pilot chart from dynamic text

### PR DESCRIPTION
**Pull request by mlbrgl**
https://user-images.githubusercontent.com/13406362/120012137-6949f980-bfdf-11eb-9b0c-e6be85e751af.mp4

This is to support the [experiment described in Pilot chart from dynamic text](https://www.notion.so/owid/Pilot-chart-from-dynamic-text-580a7d0ad84546d3b83499d3581269f0#f69fd796d1d2451587a635ff462f1b14).

**[Live demo](https://www.notion.so/owid/Pilot-chart-from-dynamic-text-580a7d0ad84546d3b83499d3581269f0#f9c6600b81cc4c29a82ddcf9335f6102)**

**What this does:**
- inject data in {{DataValue}} template strings found in content, on the server (e.g. `{{DataValue variableId:146684 entityId:355 year:2017 | %value percent in %year }}` becomes "8.8 percent in 2017")
- lazily hydrates on the client to add interactivity on hover, highlighting the relevant year on the line chart in the next column.

**Know limitations and issues:** 
- [Data in text](https://www.notion.so/owid/Data-in-text-b47f4e7018cd46dcb276d5fadeefe416#f0af7f96b5fb4e4b84db0f90ea654bd8)
- [Pilot chart from dynamic text](https://www.notion.so/owid/Pilot-chart-from-dynamic-text-580a7d0ad84546d3b83499d3581269f0#f1867dfc1b294fa2b37631606ad4aaed)

### CODE REVIEW

**Comment by danielgavrilov** [comment]
@mlbrgl Do you need a review on this?

**Comment by mlbrgl** [comment]
@danielgavrilov yes that'd be good, thanks for offering!

**Comment by danielgavrilov** [comment]
I won’t get to reviewing it this week :( if it’s blocking you someone else could review?

**Comment by mlbrgl** [comment]
thanks for following up on this. Let's reassess next week, I'm in no rush really, all good.

File: site/multiembedder/MultiEmbedder.tsx [hunk 1]
```diff
@@ -52,6 +55,7 @@ class MultiEmbedder {
     private figuresObserver: IntersectionObserver | undefined
     selection: SelectionArray = new SelectionArray()
     graphersAndExplorersToUpdate: Set<SelectionArray> = new Set()
+    grapherInstances: Map<Element, Grapher> = new Map()
```
Comment (danielgavrilov) [inline site/multiembedder/MultiEmbedder.tsx:58]: This seems unused?

File: site/AnnotatingDataValue.tsx [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+                className={isInteractive ? "interactive" : ""}
+            >
+                <DataValue label={label}></DataValue>
+                {isInteractive ? <FontAwesomeIcon icon={faChartLine} /> : null}
+            </span>
+        </span>
+    )
+}
+
+export function runAnnotatingDataValue(
```
Comment (danielgavrilov) [inline site/AnnotatingDataValue.tsx:56]: Maybe rename to `hydrateAnnotatingDataValue`? That's the only thing it does right?

File: site/tsconfig.json [hunk 3]
```diff
@@ -6,6 +6,7 @@
     },
     "references": [
         { "path": "../clientUtils" },
+        { "path": "../db" },
```
Comment (danielgavrilov) [inline site/tsconfig.json:9]: Is still still necessary? I didn't notice any imports from `db`

File: baker/formatting.tsx [hunk 4]
```diff
@@ -40,6 +47,48 @@ export const extractFormattingOptions = (html: string): FormattingOptions => {
...
+export const dataValueRegex = new RegExp(
+    `{{\\s*${DATA_VALUE}\\s*(.+?)\\s*}}`,
+    "g"
+)
+
+export const extractDataValuesConfiguration = async (
+    html: string
+): Promise<Map<string, DataValueConfiguration>> => {
+    const dataValueSeparator = /\s*\|\s*/
+    const dataValuesConfigurations = new Map()
```
Comment (danielgavrilov) [inline baker/formatting.tsx:62]: nit: maybe an explicit type for `Map`?

File: db/model/Variable.ts [hunk 5]
```diff
@@ -181,3 +182,39 @@ export async function writeVariableCSV(
...
+    year,
+}: DataValueQueryArgs) => {
+    if (!variableId || !entityId) return
+
+    const queryStart = `
+        SELECT value, year, variables.unit as unit, entities.name as entityName FROM data_values
+        JOIN entities on entities.id = data_values.entityId
+        JOIN variables on variables.id = data_values.variableId
+        WHERE entities.id = ?
+        AND variables.id= ?`
```
Comment (danielgavrilov) [inline db/model/Variable.ts:198]: If we want the value formatting to match the way the Grapher formats values, we'll also need to consider the `variables.display` configuration.  

E.g. `variables.unit` and `variables.shortUnit` columns tend to only be defined in exceptional cases, when the imported dataset has a unit defined at import time (which only happens for bulk imports).

Mostly, the units are defined in `variables.display->"$.unit"`. This is what the authors have control over through the OWID Admin interface, and the Grapher will use a `display` property, if available, and only then default to the `variables.unit` field. 

The `display` JSON field on the `variables` table has this schema: https://github.com/owid/owid-grapher/blob/9bc2033bbdc0e834d0d8589a080232d602e6f84e/grapher/core/LegacyVariableCode.ts#L27-L41

There are some additional transformations happening:
- `conversionFactor: number` multiplies the value, e.g. if the original imported dataset has a value `1000 grams`, you can set `{ conversionFactor: 0.001, unit: "kg" }` in order to turn it into `1 kg`. 
- `yearIsDay: boolean` means the "year" actually needs to be formatted as a day. The year value is actually an offset (in days) from `zeroDay: string`. 
- `numDecimalPlaces` would round the value to the specified decimal places

If this is just a very limited test, we can probably get away without these formatting issues.

Comment (danielgavrilov) [reply db/model/Variable.ts:198]: Maybe one easy extension is to pass the value through `formatValue()`, which implements at least some of these transformations: https://github.com/owid/owid-grapher/blob/95504e57f04a3e9dd9e82c78ec59475814f752bc/clientUtils/formatValue.ts#L16-L81

**Review by danielgavrilov — APPROVED** [review]
The functionality this offers is very cool!

The implementation is good for a very limited test, but we need to implement a proper, chart-independent way to pass annotations if we start extending it. 

Wiping the chart modifications on hover also isn't great, but as long as that's the desired behaviour for the test, it's fine.

File: site/tsconfig.json (hunk 3 above)
Comment (mlbrgl) [reply site/tsconfig.json:9]: nope, thanks!

File: db/model/Variable.ts (hunk 5 above)
Comment (mlbrgl) [reply db/model/Variable.ts:201]: @danielgavrilov thanks for the background, that's very helpful. I gave `formatValue` a go but it didn't go far enough given what options I could pass it at this stage. 

I'll set formatting aside for now, as the immediate experiment focuses on piloting charts. The "pulling data from the db" part is more incidental at this point and can be manual.

**Comment by mlbrgl** [comment]
> we need to implement a proper, chart-independent way to pass annotations if we start extending it.

oh yes, totally agree!

> Wiping the chart modifications on hover also isn't great

as opposed to click?

**Pull request merged**